staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -24,19 +24,273 @@ import (
+	if q.logger != nil {
+		q.log("test: Called Get()")
+	}
+	return q.delayingType.Get()
+}
+func (q *testDelayingType) Done(item interface{}) {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called Done(%v)", item))
+	}
+	q.delayingType.Done(item)
+}
+func (q *testDelayingType) ShutDown() {
+	if q.logger != nil {
+		q.log("test: Called Shutdown()")
+	}
+	q.delayingType.ShutDown()
+}
+func (q *testDelayingType) ShutDownWithDrain() {
+	if q.logger != nil {
+		q.log("test: Called ShutdownWithDrain()")
+	}
+	q.delayingType.ShutDownWithDrain()
+}
+func (q *testDelayingType) Add(item interface{}) {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called Add(%v)", item))
+	}
+	q.delayingType.Add(item)
+}
+func (q *testDelayingType) AddAfter(item interface{}, duration time.Duration) {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called AddAfter(%v, %s)", item, duration))
+	}
+	q.delayingType.AddAfter(item, duration)
+}
+func (q *testDelayingType) AddWithOptions(item interface{}, opts DelayingOptions) {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called AddWithOptions(%v, %s)", item, opts))
+	}
+	q.delayingType.AddWithOptions(item, opts)
+}
+func (q *testDelayingType) ForgetWaiting(item interface{}) {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called ForgetWaiting(%v)", item))
+	}
+	q.delayingType.ForgetWaiting(item)
+}
+
+// IsWaiting is a helper method return an answer if an item is in the 'waiting' queue
+// or not.  It returns a pointer and nil represents "no result".
+// Note: Please use IsQueued() to test if an item is in the 'active' queue.
+func (q *testDelayingType) IsWaiting(item interface{}) *bool {
+	if q.logger != nil {
+		q.log(fmt.Sprintf("test: Called IsWaiting(%v)", item))
+	}
+
+	if q.ShuttingDown() {
+		return nil
+	}
+
+	// construct a message request waiting status of item
+	cb := make(chan int)
+	waitEntry := &waitFor{
+		action:   waitForActionExists,
+		data:     item,
+		returnCh: cb,
+	}
+
+	// Send the entry to the waitingLoop...
+	select {
+	case <-q.stopCh:
+		return nil
+	case q.waitingForAddCh <- waitEntry:
+		// wait for call-back if we asked for one...
+		select {
+		case <-q.stopCh:
+			return nil
+		case result := <-cb:
+			ret := (result == 1)
+			return &ret
+		}
+	}
+}
+
+// LenWaiting returns the length of the 'waiting' queue.
+// It returns a pointer to an int and nil represents "no result".
+// Note: Please use Len() to get the length of the 'active' queue.
+func (q *testDelayingType) LenWaiting() *int {
+	if q.logger != nil {
+		q.log("test: Called LenWaiting()")
+	}
+
+	if q.ShuttingDown() {
+		return nil
+	}
+
+	// construct a message request waiting status of item
+	cb := make(chan int)
+	waitEntry := &waitFor{
+		action:   waitForActionLength,
+		returnCh: cb,
+	}
+
+	// Send the entry to the waitingLoop...
+	select {
+	case <-q.stopCh:
+		return nil
+	case q.waitingForAddCh <- waitEntry:
+		select {
+		case <-q.stopCh:
+			return nil
+		case result := <-cb:
+			return &result
+		}
```

Comment:
nit for testing purposes, we should have reasonable timeouts in all of these blocking calls. If potentially something bad happens on the other side, we will be blocking here indefinitely

Reply:
or until the test times out?  That's what I usually see happen.  I don't want to add a timeout for these test messages when the client has no such timeout.  If I was to introduce a timeout for this then I think I would need to add an error return (there presently isn't one).  I would suggest we do that in another PR because this one is already a beast that I'm begging to wonder if will ever get merged.

Reply:
This depend how you run the tests, no? That is how big (if any) you specify the timeout to be?

It shouldn't be that hard. We are already passing the T struck to the constructor. So we could just remember it and call the fail internally for some reasonable timeout?